File: chapter5/chapt5_ws/src/demo_python_tf/demo_python_tf/static_tf_broadcaster.py
import rclpy
from rclpy.node import Node
from tf2_ros import StaticTransformBroadcaster  # 静态坐标发布器
from geometry_msgs.msg import TransformStamped  # 消息接口
from tf_transformations import quaternion_from_euler  # 欧拉角转四元数
import math  # 角度转弧度

"""
pip3 install --upgrade transforms3d
"""
"""
ros2 topic list

ros2 topic echo /tf_static

ros2 topic info /tf_static -v

ros2 run tf2_ros tf2_echo base_link camera_link
"""

class StaticTFBroadcaster(Node):
    def __init__(self, node_name):
        super().__init__(node_name)
        self._static_broadcaster = StaticTransformBroadcaster(self)  # self 就是节点本身
        self.publish_static_tf()

    def publish_static_tf(self):  # 发布静态坐标: 从 base_link 到 camera_link
        transform = TransformStamped()
        transform.header.stamp = self.get_clock().now().to_msg()  # 时间戳
        transform.header.frame_id = "base_link"  # 父节点坐标系
        transform.child_frame_id = "camera_link"  # 子节点坐标系
        
        transform.transform.translation.x = 0.5
        transform.transform.translation.y = 0.3
        transform.transform.translation.z = 0.6
        
        """
            math.radians() 角度转弧度
            quaternion_from_euler 欧拉角转四元数 顺序为：[x, y, z, w]
        """
        q = quaternion_from_euler(math.radians(180), 0, 0)
        
        # 旋转部分进行赋值
        transform.transform.rotation.x = q[0]
        transform.transform.rotation.y = q[1]
        transform.transform.rotation.z = q[2]
        transform.transform.rotation.w = q[3]
        
        # 发布静态坐标关系
        self._static_broadcaster.sendTransform(transform)
        self.get_logger().info(f"发布静态TF：{transform}")
        
def main():
    rclpy.init()
    node = StaticTFBroadcaster("static_tf_broadcaster")
    rclpy.spin(node)
    rclpy.shutdown()

File: chapter5/chapt5_ws/src/demo_cpp_tf/src/static_tf_broadcaster.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"  // 提供消息接口
#include "tf2/LinearMath/Quaternion.hpp"  // 提供 tf2::Quaternion 类
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"  // 消息类型转换函数
#include "tf2_ros/static_transform_broadcaster.hpp"  // 静态坐标广播器

/*
ros2 topic list

ros2 topic echo /tf_static

ros2 topic info /tf_static -v

ros2 run tf2_ros tf2_echo map target_point
*/

class StaticTFBroadcaster: public rclcpp::Node
{
private:
    std::shared_ptr<tf2_ros::StaticTransformBroadcaster> _broadcaster;

public:
    StaticTFBroadcaster(std::string node_name): Node(node_name)
    {
        _broadcaster = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);
        this->publish_tf();
    }

    void publish_tf()
    {
        geometry_msgs::msg::TransformStamped transform;
        transform.header.stamp = this->get_clock()->now();
        transform.header.frame_id = "map";
        transform.child_frame_id = "target_point";

        transform.transform.translation.x = 5.0;  // 单位是 米
        transform.transform.translation.y = 3.0;
        transform.transform.translation.z = 0.0;
        
        tf2::Quaternion q;
        q.setRPY(0.0, 0.0, 60.0*M_PI/180.0);  // 60.0*M_PI/180.0  弧度转角度【*M_PI】，再转弧度【/180.0】

        transform.transform.rotation = tf2::toMsg(q);

        this->_broadcaster->sendTransform(transform);
    }
};

int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<StaticTFBroadcaster>("static_tf_broadcaster");
    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;
}